### PR TITLE
feat: Revamp navigation bar

### LINKED ISSUE #4428: Revamp Navigation Bar in Shows for Enhanced Usability
# Current Behavior

The navigation bar on shows currently displays the record name. Occasionally, incorrect object icons are displayed, as observed on `People` shows.

CleanShot 2024-03-12 at 16 14 33

# Desired Behavior

We aim to revamp the navigation bar on Shows to:

-  Enable navigation to the Previous or Next record within an Index View.
-  Indicate the index view from which the show was accessed.
-  Provide an option to return to the corresponding index view.
-  Display the total number of records within a view.

When a show page is accessed from a favorite or non-specific view, the "default index" name should be used as the view title (key=index), for example, `All Companies`, and the previous/next should navigate within this default index.

[Design](https://github.com/twentyhq/twenty/assets/19412894/e74039f7-7d73-4d4a-8c0d-e5a2e72e063a)

# Figma

### Component

https://www.figma.com/file/xt8O9mFeLl46C5InWwoMrN/Twenty?type=design&node-id=170-7270&mode=design&t=zAhGCjH831uYCsjq-11

### Prototype

https://www.figma.com/file/xt8O9mFeLl46C5InWwoMrN/Twenty?type=design&node-id=5910-72411&mode=design&t=zAhGCjH831uYCsjq-11

### PR DESCRIPTION
**Pull request by AdityaPimpalkar**
closes: #4428

Testing for fetchMoreRecords is pending, along with component tests

### CODE REVIEW

**Comment by FelixMalfait** [comment]
Pass view id to showpage on Kanban

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  const [currentRecordIndex, setCurrentRecordIndex] = useState(0);
+  const [hasPreviousRecord, setHasPreviousRecord] = useState(false);
+  const [hasNextRecord, setHasNextRecord] = useState(false);
+  const [objectRecords, setObjectRecords] = useState<ObjectRecord[]>([]);
+  const [totalRecords, setTotalRecords] = useState(0);
+  const [isLoadedRecords, setIsLoadedRecords] = useState(false);
+  const [isLoadingPagination, setIsLoadingPagination] = useState(true);
+  const [hasNextPage, setHasNextPage] = useState<boolean | undefined>();
+
+  const [viewName, setViewName] = useState('');
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:45]: I think it would have been great to have a shared hook like useAppPageInfo or something like that (not great naming but I don't have a better idea), so that this can be re-used in the commandBar (we'll want to adapt actions in the commandBar based on the selected view and selected object)

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    if (
+      !isLoadingPagination &&
+      objectRecords.length > 0 &&
+      isDefined(hasNextPage)
+    ) {
+      const recordIndex = objectRecords.findIndex(
+        (rec) => rec.id === objectRecordId,
+      );
+      if (recordIndex < 0 && hasNextPage) {
+        fetchMoreRecords();
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:132]: Need to change the mechanism to do two queries for previous/past results ; or maybe even better if it works, one query with an offset (can be adapted based on position) to get in a single query roughly half of the records before and half of the records after

Comment (FelixMalfait) [reply packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:132]: Note one thing we could have done to get the exact position in the list without passing it as a state from the previous state, would have been to convert the sort criteria into an additional (e.g. if it's sortBy createdAt, then we add createdAt < currentRecord.createdAt to the current record filters). It would have been perfect, but it might be a bit too painful to implement because "greater/less than" filters are probably not implemented for all field types on the backend.

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    objectMetadataItem,
+    mapViewSortsToSorts(view?.viewSorts ?? [], sortDefinitions),
+  );
+
+  const { fetchMoreRecords } = useFindManyRecords({
+    filter,
+    orderBy,
+    skip: isLoadedRecords,
+    objectNameSingular,
+    recordGqlFields: generateDepthOneRecordGqlFields({
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:111]: Should we just fetch the ids if it's all we care about? no?

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+    if (nextRecordIsLast === true && hasNextPage === true) {
+      setIsLoadedRecords(false);
+      fetchMoreRecords();
+    }
+    setHasNextRecord(nextIndex < objectRecords.length);
+  }, [
+    hasNextPage,
+    currentRecordIndex,
+    objectRecords,
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:173]: objectRecords feels to generic, if you fetch Ids only maybe you can also come up with a name that speaks more about what you're doing (e.g. orderedListOfObjectIds or something better you can come up with!)

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+  const objectNameSingular = propsObjectNameSingular || paramObjectNameSingular;
+  const objectRecordId = propsObjectRecordId || paramObjectRecordId;
+
+  if (!objectNameSingular || !objectRecordId) {
+    throw new Error('Object name or Record id is not defined');
+  }
+
+  const { objectMetadataItem } = useObjectMetadataItem({ objectNameSingular });
+
+  const { objectNamePlural } = useObjectNamePluralFromSingular({
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:56]: not important but you already have it easily with `objectMetadataItem.namePlural` so maybe no need to use this method which is more costly?

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+      }`,
+    );
+    setCurrentRecordIndex(previousIndex);
+  };
+
+  const navigateToNextRecord = () => {
+    const nextIndex = currentRecordIndex + 1;
+    const nextRecord = objectRecords[nextIndex];
+    navigate(
+      `/object/${objectNameSingular}/${nextRecord.id}${
```
[the remark this answers is not in the document]
Comment (FelixMalfait) [reply packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:196]: I know we might have this bad pattern elsewhere already (and I've probably done it too!), but since you're introducing this quite a lot of times in this PR, it would be great if we can have something a bit more elegant

**Review by FelixMalfait — COMMENTED** [review]
I made a lot of small comments but overall it is great! Thanks!!!

File: packages/twenty-front/src/modules/object-record/hooks/useFetchAllRecordIds.ts [hunk 7]
```diff
@@ -37,7 +37,7 @@ export const useFetchAllRecordIds = <T>({
     const firstQueryResult =
       findManyRecordsDataResult?.data?.[objectMetadataItem.namePlural];
 
-    const totalCount = firstQueryResult?.totalCount ?? 1;
+    const totalCount = firstQueryResult?.totalCount ?? 0;
```
Comment (lucasbordeau) [inline packages/twenty-front/src/modules/object-record/hooks/useFetchAllRecordIds.ts:40]: @thomtrp

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts (hunk 4 above)
Comment (lucasbordeau) [reply packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:174]: We need to fetch records

File: packages/twenty-front/src/modules/object-record/record-field/states/recordPositionInternalState.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import { createState } from 'twenty-ui';
+
+export const recordPositionInternalState = createState<number | null>({
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/record-field/states/recordPositionInternalState.ts:3]: Position is confusing can we rename to rank (or rankInView)

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useReasd.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import { RecordGqlRefEdge } from '@/object-record/cache/types/RecordGqlRefEdge';
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useReasd.ts:1]: Strange file name?

File: packages/twenty-front/src/modules/object-record/record-table/record-table-body/components/RecordTableBodyEffect.tsx [hunk 10]
```diff
@@ -79,6 +84,43 @@ export const RecordTableBodyEffect = () => {
...
+      const isRecordAlreadyFetched = records.some(
+        (record) => record.id === lastShowPageRecordId,
+      );
+
+      if (isRecordAlreadyFetched) {
+        const recordPosition = records.findIndex(
+          (record) => record.id === lastShowPageRecordId,
+        );
+
+        const positionInPx = recordPosition * 32;
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/object-record/record-table/record-table-body/components/RecordTableBodyEffect.tsx:104]: use row height?

File: packages/twenty-front/src/modules/object-record/record-field/meta-types/display/components/ChipFieldDisplay.tsx [hunk 11]
```diff
@@ -1,23 +1,21 @@
...
-
-  return (
-    <EntityChip
-      entityId={recordValue.id}
-      name={recordChipData.name as any}
-      avatarType={recordChipData.avatarType}
-      avatarUrl={recordChipData.avatarUrl ?? ''}
-      linkToEntity={recordChipData.linkToShowPage}
+  return isLabelIdentifier ? (
+    <RecordIndexRecordChip
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-field/meta-types/display/components/ChipFieldDisplay.tsx:14]: Not sure about this naming RecordIndexRecordChip. should be RecordIdentifierChip?

File: packages/twenty-front/src/modules/object-record/record-field/meta-types/display/utils/isFieldChipDisplay.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { FieldMetadataItem } from '@/object-metadata/types/FieldMetadataItem';
+import { isFieldFullName } from '@/object-record/record-field/types/guards/isFieldFullName';
+import { isFieldNumber } from '@/object-record/record-field/types/guards/isFieldNumber';
+import { isFieldText } from '@/object-record/record-field/types/guards/isFieldText';
+
+export const isFieldChipDisplay = (
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-field/meta-types/display/utils/isFieldChipDisplay.ts:6]: I would rename this to isFieldIdentifierDisplay

File: packages/twenty-front/src/modules/object-record/record-field/states/lastShowPageRecordId.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+import { createState } from 'twenty-ui';
+
+export const lastShowPageRecordIdState = createState<string | null>({
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-field/states/lastShowPageRecordId.ts:3]: this looks to be leaving on its own, I don't get why it's in "record-field" module

Comment (charlesBochet) [reply packages/twenty-front/src/modules/object-record/record-field/states/lastShowPageRecordId.ts:3]: i would expect this to be in a view-navigation module or something like this

File: packages/twenty-front/src/modules/object-record/record-field/states/recordPositionInternalState.ts (hunk 8 above)
Comment (charlesBochet) [reply packages/twenty-front/src/modules/object-record/record-field/states/recordPositionInternalState.ts:3]: agree + same comment as aboved

File: packages/twenty-front/src/modules/object-record/record-index/components/RecordIndexContainer.tsx [hunk 14]
```diff
@@ -108,6 +124,63 @@ export const RecordIndexContainer = ({
...
+
+  const { records: views } = usePrefetchedData<View>(PrefetchKey.AllViews);
+
+  const currentViewId = useRecoilValue(
+    currentViewIdComponentState({
+      scopeId: recordIndexId,
+    }),
+  );
+
+  const view = findView({
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-index/components/RecordIndexContainer.tsx:137]: can't we use useGetCurrentView here? also I'm worried about re-renderings here, shoudn't be this either be in an effect, computing a state, either using recoilCallback?

File: packages/twenty-front/src/modules/object-record/record-index/components/RecordIndexContainer.tsx [hunk 15]
```diff
@@ -153,41 +226,46 @@ export const RecordIndexContainer = ({
...
-            />
-            <RecordIndexBoardDataLoaderEffect
-              objectNameSingular={objectNameSingular}
-              recordBoardId={recordIndexId}
-            />
-          </StyledContainerWithPadding>
-        )}
+        <RecordIndexEventContext.Provider
+          value={{
+            onIndexIdentifierClick: handleIndexIdentifierClick,
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-index/components/RecordIndexContainer.tsx:231]: nice!

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useFindRecordCursorFromFindManyCacheRootQuery.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
+import { RecordGqlRefEdge } from '@/object-record/cache/types/RecordGqlRefEdge';
+import { useApolloClient } from '@apollo/client';
+import { createApolloStoreFieldName } from '~/utils/createApolloStoreFieldName';
+
+export const useFindRecordCursorFromFindManyCacheRootQuery = ({
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useFindRecordCursorFromFindManyCacheRootQuery.ts:5]: couldn't we have two queries with same fieldVariables? I think we have no other choice than to pass the cursor in the url / store it in a recoil state

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordIdsFromFindManyCacheRootQuery.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import { RecordGqlRefEdge } from '@/object-record/cache/types/RecordGqlRefEdge';
+import { useApolloClient } from '@apollo/client';
+import { createApolloStoreFieldName } from '~/utils/createApolloStoreFieldName';
+
+export const useRecordIdsFromFindManyCacheRootQuery = ({
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordIdsFromFindManyCacheRootQuery.ts:5]: regarding the comment above, here it would be find if fieldVariables was also containing the cursor

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+  objectMetadataItemId,
+  views,
+}: {
+  viewId: string | null;
+  objectMetadataItemId: string;
+  views: View[];
+}) => {
+  if (!viewId) {
+    return views.find(
+      (view: any) =>
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:39]: any could be typed here

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+
+  const recordBeforeCursor = pageInfoBefore?.endCursor;
+  const recordAfterCursor = pageInfoAfter?.endCursor;
+
+  const navigateToPreviousRecord = () => {
+    navigate(
+      `/object/${objectNameSingular}/${recordBefore.id}${
+        viewIdQueryParam ? `?view=${viewIdQueryParam}` : ''
+      }`,
+      {
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:195]: do we use it?

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+    ) as View;
+  } else {
+    return views.find(
+      (view: any) =>
+        view?.id === viewId && view?.objectMetadataId === objectMetadataItemId,
+    ) as View;
+  }
+};
+
+export const useRecordShowPagePagination = (
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-show/hooks/useRecordShowPagePagination.ts:50]: I feel this could be in a separate module "navigate-in-view" or something like this and could be re-used in other places later

File: packages/twenty-front/src/modules/object-record/record-table/components/RecordTableInternalEffect.tsx [hunk 21]
```diff
@@ -33,7 +33,7 @@ export const RecordTableInternalEffect = ({
     callback: () => {
       leaveTableFocus();
     },
-    mode: ClickOutsideMode.comparePixels,
+    mode: ClickOutsideMode.compareHTMLRef,
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/object-record/record-table/components/RecordTableInternalEffect.tsx:36]: why this change? don't we have an issue with portals? (filter dropdown, field input) ?

**Review by ijreilly — APPROVED** [review]

File: packages/twenty-front/src/modules/object-record/record-show/hooks/useFindRecordCursorFromFindManyCacheRootQuery.ts (hunk 16 above)
Comment (lucasbordeau) [reply packages/twenty-front/src/modules/object-record/record-show/hooks/useFindRecordCursorFromFindManyCacheRootQuery.ts:5]: I would use a recoilState

File: packages/twenty-front/src/modules/object-record/record-table/components/RecordTableInternalEffect.tsx (hunk 21 above)
Comment (lucasbordeau) [reply packages/twenty-front/src/modules/object-record/record-table/components/RecordTableInternalEffect.tsx:36]: I changed it in a recent PR but this introduced the bug of double click to open
